File: Toggly.FeatureManagement.NET/Toggly.FeatureManagement/Filters/ContextPropertyEvaluator.cs
using Microsoft.FeatureManagement;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using Toggly.FeatureManagement.Context;
using Toggly.FeatureManagement.Data;

namespace Toggly.FeatureManagement.Filters
{
    internal static class ContextPropertyEvaluator
    {
        private const string FilterName = "ContextProperty";

        private static readonly HashSet<string> EqualityOperators = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "eq", "neq" };
        private static readonly HashSet<string> ComparisonOperators = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "gt", "gte", "lt", "lte" };
        private static readonly HashSet<string> InOperators = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "in" };
        private static readonly HashSet<string> ContainsOperators = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "contains" };

        public static bool IsContextPropertyFilter(FeatureFilter? filter) =>
            filter != null && string.Equals(filter.Name, FilterName, StringComparison.OrdinalIgnoreCase);

        public static IReadOnlyList<FeatureFilter> GetEntityFilters(FeatureDefinitionModel definition) =>
            definition.Filters.Where(IsContextPropertyFilter).ToList();

        public static IReadOnlyList<FeatureFilter> GetUserFilters(FeatureDefinitionModel definition) =>
            definition.Filters.Where(f => !IsContextPropertyFilter(f)).ToList();

        public static bool HasEntityFilters(FeatureDefinitionModel definition) =>
            definition.Filters.Any(IsContextPropertyFilter);

        public static bool HasUserFilters(FeatureDefinitionModel definition) =>
            definition.Filters.Any(f => !IsContextPropertyFilter(f));

        public static bool EvaluateEntityFilters(
            FeatureDefinitionModel definition,
            TogglyEntityContext entity)
        {
            var filters = GetEntityFilters(definition);
            if (filters.Count == 0)
                return false;

            var results = filters.Select(filter => EvaluateSingleFilter(filter, entity)).ToList();
            var all = IsAllRequirement(definition.ContextRequirementType ?? definition.RequirementType);
            return all ? results.All(r => r) : results.Any(r => r);
        }

        private static bool IsAllRequirement(RequirementType requirementType) =>
            requirementType == RequirementType.All;

        private static bool EvaluateSingleFilter(FeatureFilter filter, TogglyEntityContext entity)
        {
            var parameters = filter.Parameters ?? new Dictionary<string, string>();
            parameters.TryGetValue("Property", out var propertyName);
            parameters.TryGetValue("Operator", out var op);
            parameters.TryGetValue("Value", out var expectedValue);
            parameters.TryGetValue("ValueType", out var valueType);

            if (string.IsNullOrWhiteSpace(propertyName) || string.IsNullOrWhiteSpace(op) || expectedValue == null)
                return false;

            op = op.ToLowerInvariant();
            valueType = (valueType ?? "string").ToLowerInvariant();

            if (!entity.Attributes.TryGetValue(propertyName, out var actualValue))
                return false;

            return Compare(actualValue, op, expectedValue, valueType);
        }

        private static bool Compare(object? actual, string op, string expected, string valueType)
        {
            if (EqualityOperators.Contains(op))
                return CompareEquality(actual, expected, op == "eq");

            if (ComparisonOperators.Contains(op))
                return CompareOrdered(actual, expected, valueType, op);

            if (InOperators.Contains(op))
                return CompareIn(actual, expected);

            if (ContainsOperators.Contains(op))
                return CompareContains(actual, expected, valueType);

            return false;
        }

        private static bool CompareEquality(object? actual, string expected, bool shouldEqual)
        {
            var actualString = Convert.ToString(actual, CultureInfo.InvariantCulture) ?? string.Empty;
            var equal = string.Equals(actualString, expected, StringComparison.OrdinalIgnoreCase);
            return shouldEqual ? equal : !equal;
        }

        private static bool CompareOrdered(object? actual, string expected, string valueType, string op)
        {
            if (valueType == "datetime")
            {
                if (!TryParseDateTime(actual, out var actualDate) || !TryParseDateTime(expected, out var expectedDate))
                    return false;

                return op switch
                {
                    "gt" => actualDate > expectedDate,
                    "gte" => actualDate >= expectedDate,
                    "lt" => actualDate < expectedDate,
                    "lte" => actualDate <= expectedDate,
                    _ => false
                };
            }

            if (valueType == "number" && TryParseDecimal(actual, out var actualNumber) && decimal.TryParse(expected, NumberStyles.Any, CultureInfo.InvariantCulture, out var expectedNumber))
            {
                return op switch
                {
                    "gt" => actualNumber > expectedNumber,
                    "gte" => actualNumber >= expectedNumber,
                    "lt" => actualNumber < expectedNumber,
                    "lte" => actualNumber <= expectedNumber,
                    _ => false
                };
            }

            return false;
        }

        private static bool CompareIn(object? actual, string expected)
        {
            var candidates = expected
                .Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(candidate => candidate.Trim())
                .Where(candidate => candidate.Length > 0);
            var actualString = Convert.ToString(actual, CultureInfo.InvariantCulture) ?? string.Empty;
            return candidates.Any(candidate => string.Equals(candidate, actualString, StringComparison.OrdinalIgnoreCase));
        }

        private static bool CompareContains(object? actual, string expected, string valueType)
        {
            if (valueType == "string[]" && actual is IEnumerable<object> values)
                return values.Any(v => string.Equals(Convert.ToString(v, CultureInfo.InvariantCulture), expected, StringComparison.OrdinalIgnoreCase));

            var actualString = Convert.ToString(actual, CultureInfo.InvariantCulture) ?? string.Empty;
            return actualString.Contains(expected, StringComparison.OrdinalIgnoreCase);
        }

        private static bool TryParseDateTime(object? value, out DateTimeOffset result)
        {
            result = default;
            if (value is DateTimeOffset dto)
            {
                result = dto;
                return true;
            }

            if (value is DateTime dt)
            {
                if (dt.Kind == DateTimeKind.Unspecified)
                    dt = DateTime.SpecifyKind(dt, DateTimeKind.Utc);
                result = new DateTimeOffset(dt);
                return true;
            }

            var text = Convert.ToString(value, CultureInfo.InvariantCulture);
            return !string.IsNullOrWhiteSpace(text) && DateTimeOffset.TryParse(text, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out result);
        }

        private static bool TryParseDecimal(object? value, out decimal result)
        {
            if (value is decimal d)
            {
                result = d;
                return true;
            }

            var text = Convert.ToString(value, CultureInfo.InvariantCulture);
            return decimal.TryParse(text, NumberStyles.Any, CultureInfo.InvariantCulture, out result);
        }
    }
}

File: Toggly.FeatureManagement.Java/toggly-core/src/main/java/io/toggly/core/eval/ContextPropertyEvaluator.java
package io.toggly.core.eval;

import io.toggly.core.context.TogglyEntityContext;
import io.toggly.core.model.FeatureDefinition;
import io.toggly.core.model.FeatureFilter;
import io.toggly.core.model.FeatureRequirement;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Locale;

/**
 * Evaluates ContextProperty filters against an entity. Unknown operators fail closed.
 */
public final class ContextPropertyEvaluator implements FilterEvaluator {

    public static final String FILTER_NAME = "ContextProperty";
    public static final ContextPropertyEvaluator INSTANCE = new ContextPropertyEvaluator();

    private ContextPropertyEvaluator() {}

    public static boolean isContextPropertyFilter(FeatureFilter filter) {
        return filter != null && FILTER_NAME.equalsIgnoreCase(filter.getName());
    }

    public static List<FeatureFilter> getEntityFilters(FeatureDefinition definition) {
        List<FeatureFilter> out = new ArrayList<>();
        if (definition == null || definition.getFilters() == null) {
            return out;
        }
        for (FeatureFilter filter : definition.getFilters()) {
            if (isContextPropertyFilter(filter)) {
                out.add(filter);
            }
        }
        return out;
    }

    public static List<FeatureFilter> getUserFilters(FeatureDefinition definition) {
        List<FeatureFilter> out = new ArrayList<>();
        if (definition == null || definition.getFilters() == null) {
            return out;
        }
        for (FeatureFilter filter : definition.getFilters()) {
            if (!isContextPropertyFilter(filter)) {
                out.add(filter);
            }
        }
        return out;
    }

    public static boolean evaluateEntityFilters(FeatureDefinition definition, TogglyEntityContext entity) {
        List<FeatureFilter> filters = getEntityFilters(definition);
        if (filters.isEmpty() || entity == null) {
            return false;
        }
        FeatureRequirement requirement = definition.getContextRequirementType() != null
                ? definition.getContextRequirementType()
                : definition.getRequirementType();
        boolean all = requirement == FeatureRequirement.ALL;
        if (all) {
            for (FeatureFilter filter : filters) {
                if (!evaluateSingleFilter(filter, entity)) {
                    return false;
                }
            }
            return true;
        }
        for (FeatureFilter filter : filters) {
            if (evaluateSingleFilter(filter, entity)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean evaluate(FeatureFilter filter, String featureKey, io.toggly.core.context.EvaluationContext context) {
        if (context == null || context.getEntity() == null) {
            return false;
        }
        return evaluateSingleFilter(filter, context.getEntity());
    }

    static boolean evaluateSingleFilter(FeatureFilter filter, TogglyEntityContext entity) {
        String propertyName = param(filter, "Property");
        String op = param(filter, "Operator");
        String expectedValue = param(filter, "Value");
        String valueType = param(filter, "ValueType");
        if (propertyName == null || propertyName.isBlank() || op == null || op.isBlank() || expectedValue == null) {
            return false;
        }
        op = op.toLowerCase(Locale.ROOT);
        if (valueType == null || valueType.isBlank()) {
            valueType = "string";
        } else {
            valueType = valueType.toLowerCase(Locale.ROOT);
        }
        if (!entity.containsAttribute(propertyName)) {
            return false;
        }
        return compare(entity.getAttribute(propertyName), op, expectedValue, valueType);
    }

    private static String param(FeatureFilter filter, String key) {
        Object value = filter.getParameters().get(key);
        if (value == null) {
            for (var entry : filter.getParameters().entrySet()) {
                if (key.equalsIgnoreCase(entry.getKey()) && entry.getValue() != null) {
                    return String.valueOf(entry.getValue());
                }
            }
            return null;
        }
        return String.valueOf(value);
    }

    private static boolean compare(Object actual, String op, String expected, String valueType) {
        return switch (op) {
            case "eq" -> compareEquality(actual, expected, true);
            case "neq" -> compareEquality(actual, expected, false);
            case "gt", "gte", "lt", "lte" -> compareOrdered(actual, expected, valueType, op);
            case "in" -> compareIn(actual, expected);
            case "contains" -> compareContains(actual, expected, valueType);
            default -> false;
        };
    }

    private static boolean compareEquality(Object actual, String expected, boolean shouldEqual) {
        String actualString = actual == null ? "" : String.valueOf(actual);
        boolean equal = actualString.equalsIgnoreCase(expected);
        return shouldEqual == equal;
    }

    private static boolean compareOrdered(Object actual, String expected, String valueType, String op) {
        if ("datetime".equals(valueType)) {
            OffsetDateTime actualDate = parseDateTime(actual);
            OffsetDateTime expectedDate = parseDateTime(expected);
            if (actualDate == null || expectedDate == null) {
                return false;
            }
            int cmp = actualDate.compareTo(expectedDate);
            return ordered(cmp, op);
        }
        if ("number".equals(valueType)) {
            BigDecimal actualNumber = parseDecimal(actual);
            BigDecimal expectedNumber = parseDecimal(expected);
            if (actualNumber == null || expectedNumber == null) {
                return false;
            }
            return ordered(actualNumber.compareTo(expectedNumber), op);
        }
        return false;
    }

    private static boolean ordered(int cmp, String op) {
        return switch (op) {
            case "gt" -> cmp > 0;
            case "gte" -> cmp >= 0;
            case "lt" -> cmp < 0;
            case "lte" -> cmp <= 0;
            default -> false;
        };
    }

    private static boolean compareIn(Object actual, String expected) {
        String actualString = actual == null ? "" : String.valueOf(actual);
        for (String candidate : expected.split(",")) {
            String trimmed = candidate.trim();
            if (!trimmed.isEmpty() && trimmed.equalsIgnoreCase(actualString)) {
                return true;
            }
        }
        return false;
    }

    private static boolean compareContains(Object actual, String expected, String valueType) {
        if ("string[]".equals(valueType) && actual instanceof Collection<?> values) {
            for (Object value : values) {
                if (String.valueOf(value).equalsIgnoreCase(expected)) {
                    return true;
                }
            }
            return false;
        }
        if ("string[]".equals(valueType) && actual != null && actual.getClass().isArray()) {
            int length = java.lang.reflect.Array.getLength(actual);
            for (int i = 0; i < length; i++) {
                Object value = java.lang.reflect.Array.get(actual, i);
                if (String.valueOf(value).equalsIgnoreCase(expected)) {
                    return true;
                }
            }
            return false;
        }
        String actualString = actual == null ? "" : String.valueOf(actual);
        return actualString.toLowerCase(Locale.ROOT).contains(expected.toLowerCase(Locale.ROOT));
    }

    private static OffsetDateTime parseDateTime(Object value) {
        if (value instanceof OffsetDateTime odt) {
            return odt;
        }
        if (value instanceof java.time.Instant instant) {
            return OffsetDateTime.ofInstant(instant, ZoneOffset.UTC);
        }
        if (value instanceof java.util.Date date) {
            return OffsetDateTime.ofInstant(date.toInstant(), ZoneOffset.UTC);
        }
        String text = value == null ? null : String.valueOf(value);
        if (text == null || text.isBlank()) {
            return null;
        }
        try {
            return OffsetDateTime.parse(text, DateTimeFormatter.ISO_DATE_TIME);
        } catch (DateTimeParseException ignored) {
            try {
                return java.time.LocalDateTime.parse(text).atOffset(ZoneOffset.UTC);
            } catch (DateTimeParseException ignored2) {
                try {
                    return java.time.LocalDate.parse(text).atStartOfDay().atOffset(ZoneOffset.UTC);
                } catch (DateTimeParseException e) {
                    return null;
                }
            }
        }
    }

    private static BigDecimal parseDecimal(Object value) {
        if (value instanceof BigDecimal bd) {
            return bd;
        }
        if (value instanceof Number number) {
            return new BigDecimal(number.toString());
        }
        try {
            return new BigDecimal(String.valueOf(value));
        } catch (NumberFormatException e) {
            return null;
        }
    }
}
